tests/unittests/test_ds_identify.py

```diff
@@ -1079,7 +1083,11 @@ def _print_run_output(rc, out, err, cfg, files):
     'Ec2-E24Cloud-negative': {
         'ds': 'Ec2',
         'files': {P_SYS_VENDOR: 'e24cloudyday\n'},
-    }
+    },
+    'EquinixMetal': {
+        'ds': 'EquinixMetal',
+        'mocks': [], # TODO how do I mock metadata responses?
```

Comment:
You don't.  at least not for ds-identify.  It *only* identifies via locally available data.

Reply:
I could add DS support for the models that do report "Packet" back somewhere in the DMI, but I would have to survey the available models to identify any variations in the fields and values.

Should I (can I?) remove DS support and only rely on metadata service detection?

Reply:
I'm not sure what you mean by "DS support".

Maybe some background would help you to understand what ds-identify does here.

 1. We want vendors to be able to make OS images (like https://cloud-images.ubuntu.com/) that "just work" wherever you try to run them, and an OS with cloud-init installed but not on a cloud would not do anything differently.  `datasource_list` can configure which datasources will be searched, but ideally there would be no need for such a thing.  cloud-init would just "do the right thing".
 
 2. Previously, cloud-init (in python) would walk through each datasource in datasource_list and try to get data.  That meant that boot was *always* impacted (cloud-init always ran).  On EC2, that meant doing a dhcp and checking to see if the metadata service was there.  That is obviously less than ideal.  It was slow, and meant if you booted such an image elsewhere, and there happened to be http://169.254.169.254/latest/user-data, then it would execute that code.

Now, with ds-identify if it determines that the system is not on a cloud platform, cloud-init does not run at all.  From systemd's perspective, cloud-init.target is not even enabled.  But in order to do that... we only look at *local* data.  We want those checks to be very fast, and thus far, they are.  When ds-identify finds that it is on Equinox, as told to it by DMI data, it *knows* that it will find an equinox metadata service (or... if someone is lying to it, then failure is somewhat expected.  As an example... if cpu identifies itself as x86_64, but didn't implement some of the interfaces, you'd expect that a program might fail).

Reply:
Are there examples of providers that have no local representation (no guarantee of identifiable DMI ids)?

The presence of packet.net (or equinix) in metadata.platformequinix.com/2009-04-04/iqn is the only approach that will work for a majority of our infrastructure (that I am aware of).

/cc @mmlb @dustinmiller1337 @pereztr5

Reply:
@smoser Is there a possible alternative to dmi data? We can likely go that route for our own machines, but should not be an expectation for making use of tinkerbell. We _can_ control kernel boot args very easily, can we have cloud-init also check there?

Reply:
oh wait this would apply at runtime not install-time :man_facepalming: so we don't have as much control over kernel args :/

Reply:
@smoser Tinkerbell has a descendent metadata service and deployment architecture, but in that environment users bring their own hardware and DMI stamping may not be possible.

Related issue: https://github.com/tinkerbell/cluster-api-provider-tink/issues/6

Reply:
I'm leaving Tinkerbell concerns out of this PR, but I was hopeful that we could leverage this PR in support of https://tinkerbell.org/ environments later.

Reply:
> Are there examples of providers that have no local representation (no guarantee of identifiable DMI ids)?

MAAS. It sounds like MAAS is something very similar to what you're working on.
The way MAAS works is:
 * network booted environment sends cmdline with 'ci.ds=MAAS'

   ds-identify generically reads the ci.ds kernel parameter to be declaring
the datasource to use.

 * installed system declares the datasource_list to have only MAAS in it.

   during the install, maas writes a cloud-init config file to the target system. that declares 'datasource_list' to just have MAAS and ds-identify reespects that.